src/category_theory/idempotents/karoubi.lean

```diff
@@ -0,0 +1,260 @@
+Authors: Joël Riou
+-/
+
+import category_theory.idempotents.basic
+import category_theory.preadditive.additive_functor
+import category_theory.equivalence
+
+/-!
+# The Karoubi envelope of a category
+
+In this file, we define the Karoubi envelope `karoubi C` of a category `C`.
+
+## Main constructions and definitions
+
+- `karoubi C` is the Karoubi envelope of a category `C`.
+It is preadditive when `C` is preadditive.
+- `karoubi_is_idempotent_complete C` says that `karoubi C` is idempotent complete.
+- `to_karoubi C : C ⥤ karoubi C` is a fully faithful functor, which is an equivalence
+(`to_karoubi_is_equivalence`) when `C` is idempotent complete.
+
+-/
+
+noncomputable theory
+
+open category_theory.category
+open category_theory.preadditive
+open category_theory.limits
+open_locale big_operators
+
+namespace category_theory
+
+variables (C : Type*) [category C]
+
+namespace idempotents
+
+/-- In a preadditive category `C`, when an object `X` decomposes as `X ≅ P ⨿ Q`, one may
+consider `P` as a direct factor of `X` and up to unique isomorphism, it is determined by the
+obvious idempotent `X ⟶ P ⟶ X` which is the projector on `P` with kernel `Q`. More generally,
+one may define a formal direct factor of an object `X : C` : it consists of an idempotent
+`p : X ⟶ X` which is thought as the "formal image" of `p`. The type `karoubi C` shall be the
+type of the objects of the karoubi enveloppe of `C`. It makes sense for any category `C`. -/
+@[nolint has_inhabited_instance]
+structure karoubi := (X : C) (p : X ⟶ X) (idempotence : p ≫ p = p)
+
+namespace karoubi
+
+variables {C}
+
+@[ext]
+lemma ext {P Q : karoubi C} (h_X : P.X = Q.X)
+  (h_p : P.p ≫ eq_to_hom h_X = eq_to_hom h_X ≫ Q.p) : P = Q :=
+begin
+  cases P,
+  cases Q,
+  dsimp at h_X h_p,
+  subst h_X,
+  simpa only [true_and, eq_self_iff_true, id_comp, eq_to_hom_refl,
+    heq_iff_eq, comp_id] using h_p,
+end
+
+/-- A morphism `P ⟶ Q` in the category `karoubi C` is a morphism in the underlying category
+`C` which satisfies a relation, which in the preadditive case, expresses that it induces a
+map between the corresponding "formal direct factors" and that it vanishes on the complement
+formal direct factor. -/
+@[ext]
+structure hom (P Q : karoubi C) := (f : P.X ⟶ Q.X) (comm : f = P.p ≫ f ≫ Q.p)
+
+instance [preadditive C] (P Q : karoubi C) : inhabited (hom P Q) :=
+⟨⟨0, by rw [zero_comp, comp_zero]⟩⟩
+
+@[simp]
+lemma hom_ext {P Q : karoubi C} {f g : hom P Q} : f = g ↔ f.f = g.f :=
+begin
+  split,
+  { intro h, rw h, },
+  { ext, }
+end
+
+lemma p_comp {P Q : karoubi C} (f : hom P Q) : P.p ≫ f.f = f.f :=
+by rw [f.comm, ← assoc, P.idempotence]
+
+lemma comp_p {P Q : karoubi C} (f : hom P Q) : f.f ≫ Q.p = f.f :=
+by rw [f.comm, assoc, assoc, Q.idempotence]
+
+lemma p_comm {P Q : karoubi C} (f : hom P Q) : P.p ≫ f.f = f.f ≫ Q.p :=
+by rw [p_comp, comp_p]
+
+lemma comp_proof {P Q R : karoubi C} (g : hom Q R) (f : hom P Q) :
```

Comment:
I don't understand this name

Reply:
Thanks for the comments. I have done the modifications. For this one "p_comm" express a kind of "commutativity" property of morphism f with the "p" that are given on P.X and Q.X. Would "p_comp_comm" be better?

Reply:
With the `reassoc` attributes from the other suggestions, this proof should be `by simp`. So maybe the entire lemma can be dropped?

Reply:
I added `p_comm` because it was convenient in some applications. For example, when working with homological complexes, I can get some equalities by starting with a syntax like `homological_complex.congr_hom (karoubi.p_comm f) i`, which would require more steps if only p_comp and comp_p were available.